### PR TITLE
Feat - Add vue-kakuyaku

### PR DESCRIPTION
**Pull request by VladislavPopovSR**
- Integrated `vue-kayuaky/core` package;
- Created separated `BasePagination` component instead;
- Corrected styles.

### CODE REVIEW

**Comment by VladislavPopovSR** [comment]
Now we have the following fetching approach:
1. Create reactive `listState` - our query params
1.1 Create fetching function e.g. if some params should be ommitted
1.2 Write watch where resetting page if necessary (e.g. when we not on the initial page and apply filters)
2. Create scope. Some key's might be conditional
3. Write and use computed variables to access scope data

**Comment by VladislavPopovSR** [comment]
Please take a look. I had this jenkins error
Снимок экрана 2024-11-15 в 00 56 30

As I understood, it caused by discrepancy between my local pnpm version and package.json `packageManager` field. 
So after many times I finally installed `9.13.1` and changed package.json version to this. Even now in spite of the successfull jenkins check it gives incompatible lockfile file.
Снимок экрана 2024-11-15 в 00 43 38
Maybe you are familiar with this? Seems an error on my side

File: src/shared/api/handle-param-scope.ts
Comment (0x009922) [inline src/shared/api/handle-param-scope.ts:1]: This isn't directly related to `shared/api`. It is just a composable which does something with `vue-kakuyaku`, which does something with promises.

File: src/shared/api/handle-param-scope.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import type { Reactive } from 'vue';
+import { computed, reactive } from 'vue';
+import { useErrorHandlers } from '@/shared/ui/composables/useErrorHandlers';
+import { useErrorRetry, useTask, wheneverRejected } from '@vue-kakuyaku/core';
+
+export function handleParamScope<T, K>(
```
Comment (0x009922) [inline src/shared/api/handle-param-scope.ts:6]: `handle` is not a conventional prefix for composables (which this function is). More over, this function doesn't set up a scope (doesn't rely on EffectScope API). It is a plain composable that sets up some reactive logic.

Consider other options:

- `useAsyncData`
- `setupAsyncData`

File: src/shared/api/schemas.ts [hunk 2]
```diff
@@ -21,12 +21,10 @@ export interface Paginated<T> {
...
-const PaginationParams = z
-  .object({
-    page: z.number(),
-    per_page: z.number(),
-  })
-  .partial();
+const PaginationParams = z.object({
+  page: z.number(),
+  per_page: z.number(),
+});
```
Comment (0x009922) [inline src/shared/api/schemas.ts:27]: I see that this schema isn't used anywhere. Therefore, it is enough to define it just as an `interface`.

File: src/shared/api/handle-param-scope.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  return reactive({
+    isLoading: computed(() => state.pending),
+    data: computed<K | null>(() => {
+      const res = state.fulfilled?.value;
+
+      if (!res) return null;
+
+      return res;
+    }),
```
Comment (0x009922) [inline src/shared/api/handle-param-scope.ts:32]:
```suggestion
    data: computed<K | null>(() => state.fulfilled?.value),
```

File: src/shared/api/handle-param-scope.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import type { Reactive } from 'vue';
+import { computed, reactive } from 'vue';
+import { useErrorHandlers } from '@/shared/ui/composables/useErrorHandlers';
+import { useErrorRetry, useTask, wheneverRejected } from '@vue-kakuyaku/core';
+
+export function handleParamScope<T, K>(
+  payload: Reactive<T>,
+  fn: (query: Reactive<T>) => Promise<K>,
```
Comment (0x009922) [inline src/shared/api/handle-param-scope.ts:8]:
```suggestion
  fn: () => Promise<K>,
```

You can include the payload in the closure naturally, without needing to introduce any arguments here:

```ts
handleParamScope(payload, http.fetchTransactions)
// vs
handleParamScope(() => http.fetchTransactions(payload))
```

File: src/shared/ui/components/BasePagination.vue
Comment (0x009922) [inline src/shared/ui/components/BasePagination.vue:1]: Please move the complicated and messy logic into a separate module, and organise it as pure functions (input -> output, no side effects, just the logic itself), and cover them with tests. This way, the dirty staff will be localised in a single place, not spoiling the rest code with complexity and mess.

This must have been an initial approach with this code. TDD is very helpful in such a case.

File: src/shared/ui/components/InstructionsTable.vue [hunk 5]
```diff
@@ -45,36 +44,50 @@ function getInstructionPayloadEntity(item: Instruction) {
...
+    });
+  } else
+    return await http.fetchInstructions({
+      ...objectOmit(params, shouldShowKind.value ? ['kind', 'transaction_status'] : ['transaction_status']),
+    });
+}
 
-watch(listState, fetchInstructions, { immediate: true });
+watch(
+  () => [listState.kind, listState.transaction_status, listState.per_page],
```
Comment (0x009922) [inline src/shared/ui/components/InstructionsTable.vue:72]: This watch will trigger even if the actual values in the array aren't changed (because the array object will be new each time this function is called).

Make three separate watch sources instead:
```suggestion
  [() => listState.kind, () => listState.transaction_status, () => listState.per_page],
```

File: src/shared/ui/components/InstructionsTable.vue [hunk 6]
```diff
@@ -45,36 +44,50 @@ function getInstructionPayloadEntity(item: Instruction) {
...
+  transaction_status: null,
+  authority: computed(() => (props.filterBy.kind === 'authority' ? props.filterBy.value.toString() : '')),
+  transaction_hash: computed(() => (props.filterBy.kind === 'transaction' ? props.filterBy.value : '')),
   kind: 'All' as ftm.TabInstructions,
+  page: 1,
+  per_page: 10,
 });
 
-const shouldShowKind = computed(() => listState.kind === 'All');
+async function fetchInstructions(params: Reactive<typeof listState>) {
```
Comment (0x009922) [inline src/shared/ui/components/InstructionsTable.vue:59]: The `Reactive` type is probably misused. `listState`'s type is already with all refs unwrapped, which is what `reactive()` returns.

Comment (0x009922) [inline src/shared/ui/components/InstructionsTable.vue:59]: I'd suggest removing this function.

Instead, compute the actual params reactively:

```ts
const params = computed<InstructionSearchParams>(() => ...)
```

and setup the scope with them directly:

```ts
useParamScope(
  () => ({ key: JSON.stringify(params.value), payload: params }),
  ({ payload }) => handleParamScope(() => http.fetchInstructions(payload))
)
```

This way it could turn out that the whole `listState` is extra.

File: src/shared/ui/components/TransactionsTable.vue
Comment (0x009922) [inline src/shared/ui/components/TransactionsTable.vue:1]: Same comments as in `InstructionsTable`.

File: src/widgets/home-page-info/HomePageInfo.vue [hunk 7]
```diff
@@ -99,8 +99,12 @@ const nodesAmount = ref(0);
 
 const { handleUnknownError } = useErrorHandlers();
 
+const isLoading = ref(false);
```
Comment (0x009922) [inline src/widgets/home-page-info/HomePageInfo.vue:102]: Replace with `usePromise` or `useTask`.

The point of this PR is to reduce this repetitive manual data-fetching related reactivity logic.

File: src/widgets/latest-blocks/LatestBlocks.vue [hunk 8]
```diff
@@ -47,33 +47,33 @@ import BaseLink from '@/shared/ui/components/BaseLink.vue';
...
+  () => {
+    return {
+      key: 'latest-blocks',
+      payload: {
+        per_page: 10,
+      },
+    };
+  },
+  ({ payload }) => handleParamScope(payload, http.fetchBlocks)
+);
```
Comment (0x009922) [inline src/widgets/latest-blocks/LatestBlocks.vue:69]: The point of setting up a scope is that you have independent virtual reactive environments for each set of parameters. Thus, if parameters change, you don't need to set up various watchers to stop current fetching and start a new one _manually_ - the scope is simply disposed (as if it was a separate component) and a new one is set up.

In this case payload is constant, thus it's pointless to set up a parametrised scope. It only introduces extra complexity.

File: src/widgets/latest-transactions/LatestTransactions.vue [hunk 9]
```diff
@@ -62,54 +62,57 @@
...
-      isLoading.value = false;
-    }
+watch(
+  total,
+  () => {
+    emit('loaded', total.value);
   },
-  { immediate: true }
+  { once: true }
 );
```
Comment (0x009922) [inline src/widgets/latest-transactions/LatestTransactions.vue:116]: A more correct and elegant way would be to put it directly into the scope where data is loaded, and use the `wheneverFulfilled` helper. 

```ts
wheneverFulfilled(state, (value) => emit('loaded', value.pagination.total_items))
```

**Review by 0x009922 — CHANGES_REQUESTED** [review]

File: src/widgets/latest-transactions/LatestTransactions.vue (hunk 9 above)
Comment (VladislavPopovSR) [reply src/widgets/latest-transactions/LatestTransactions.vue:116]: Thats great idea but I've just realized that there is no need to fetch transactions and blocks - we can use `blocks`, `txs_rejected` and `txs_accepted` fields when fetching peer status in order to get their total count

File: package.json [hunk 10]
```diff
@@ -32,8 +33,9 @@
     "zod": "^3.23.8"
   },
   "devDependencies": {
+    "@stylistic/eslint-plugin-ts": "^2.11.0",
```
Comment (0x009922) [inline package.json:36]: Why?

File: src/shared/ui/composables/usePagination.ts
Comment (0x009922) [inline src/shared/ui/composables/usePagination.ts:1]: It is still not what I meant.

Inside `usePagination`, you have computation of `segmentInfo` and of `numbers`. Make them separate functions which have nothing to do with reactivity, _just pure logic_:

```ts
function computeSegmentInfo(input: {
	totalItems?: number,
	page: number,
	reversed: boolean
	/* ... */
}): { start: number, end: number, total: number } 

function computeNumbers(input: {
	/* put your params here */
}): (number | 'ellipsis')[]  
```

Then, test these functions directly. Pure input, pure output. All corner cases.

Then, define a composable, `usePagination`, on top of it. Please take a look at how composables are usually defined in VueUse. Specifically, I don't like that you pass all parameters in a single `ComputedRef<...>`. Make them all separate, make each `MaybeRefOrGetter<T>`.

Then, on top of it, add a few high-level tests of the actual component `<BasePagination>`, which doesn't test pagination, but only how it is displayed.

File: src/shared/utils/setup-async-data.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  useErrorRetry(state, run, { interval: options?.interval ?? 5000 });
+
+  wheneverRejected(state, (err) => {
+    handleUnknownError(err);
+  });
+
+  return reactive({
+    isLoading: computed(() => state.pending),
+    data: computed<K | undefined>(() => state.fulfilled?.value),
+  });
```
Comment (0x009922) [inline src/shared/utils/setup-async-data.ts:25]: Why not just return `{ state }` directly.

File: src/widgets/latest-blocks/LatestBlocks.vue [hunk 12]
```diff
@@ -47,34 +47,14 @@ import BaseLink from '@/shared/ui/components/BaseLink.vue';
...
-import { onMounted, ref, shallowRef } from 'vue';
-import { useErrorHandlers } from '@/shared/ui/composables/useErrorHandlers';
+import { computed } from 'vue';
 import TimeStamp from '@/shared/ui/components/TimeStamp.vue';
+import { setupAsyncData } from '@/shared/utils/setup-async-data';
 
-const emit = defineEmits<{
-  loaded: [number]
-}>();
+const setup = setupAsyncData(() => http.fetchBlocks({ per_page: 10 }));
```
Comment (0x009922) [inline src/widgets/latest-blocks/LatestBlocks.vue:54]: Again, look at how composables are defined in VueUse. You usually can use them in a way like this:

```ts
const { isLoading, data: blocks } = setupAsyncData(...) 
``` 

Consider aligning your composable with what's an established standard in the ecosystem.

**Review by 0x009922 — CHANGES_REQUESTED** [review]

File: package.json (hunk 10 above)
Comment (VladislavPopovSR) [reply package.json:36]: Well DevOps said that there was a problem with `typescript-eslint/eslint-plugin` package. After I've installed the latest version I've found out that some rules are deprecated/moved to new packages. E.g. [member-delimiter-style](https://typescript-eslint.io/rules/member-delimiter-style/) rule.

File: src/shared/utils/setup-async-data.ts (hunk 11 above)
Comment (VladislavPopovSR) [reply src/shared/utils/setup-async-data.ts:25]: In order to access only `expose` field since we don't use other fields for now and every time have the same output with only two fields: `isLoading` and `data`

**Comment by VladislavPopovSR** [comment]
I think for `BasePagination.test.ts` there is no need to move data to variable and reuse it since every test has different amount of cases and different fields passed within

File: package.json (hunk 10 above)
Comment (0x009922) [reply package.json:36]: There is no longer a problem with packages in CI. Can we revert extra changes?

File: src/widgets/latest-transactions/LatestTransactions.vue [hunk 13]
```diff
@@ -62,56 +62,46 @@
...
-      isLoading.value = true;
-
-      const params = { per_page: 5, status: status.value };
+const listState = reactive({
+  per_page: 5,
+  status: null,
+});
 
-      const res = await http.fetchTransactions(objectOmit(params, status.value ? [] : ['status']));
+async function fetchTransactions(params: Reactive<typeof listState>) {
```
Comment (0x009922) [inline src/widgets/latest-transactions/LatestTransactions.vue:85]: Why do you use `Reactive` here?

File: src/pages/TransactionDetails.vue [hunk 14]
```diff
@@ -44,24 +42,24 @@ const transactionHash = computed(() => {
...
-    } finally {
-      isFetchingTransaction.value = false;
-    }
+const transactionScope = useParamScope(
+  () => {
+    return {
+      key: transactionHash.value,
+      payload: transactionHash.value,
+    };
   },
```
Comment (0x009922) [inline src/pages/TransactionDetails.vue:51]: Primitive payloads could be passed as-is (they themselves will be used as the key): ([docs](https://www.npmjs.com/package/@vue-kakuyaku/core#reactively-parametrised-scope-with-useparamscope)):

```suggestion
  transactionHash,
```

File: src/pages/DomainsList.vue [hunk 15]
```diff
@@ -113,6 +99,29 @@ const hashType = computed(() => {
 
   return 'two-line';
 });
+
+const listState = reactive({
+  page: 1,
+  per_page: 10,
+});
+
+watch([() => listState.per_page], () => {
```
Comment (0x009922) [inline src/pages/DomainsList.vue:108]:
```suggestion
watch(() => listState.per_page, () => {
```

File: src/pages/AccountDetails.vue [hunk 16]
```diff
@@ -19,55 +16,93 @@ import InstructionsTable from '@/shared/ui/components/InstructionsTable.vue';
...
+});
 
-const { width } = useWindowSize();
+const assetsScope = useParamScope(
+  () => {
+    return {
+      key: account.value?.owned_assets ? JSON.stringify(assetsListState) : '',
+      payload: assetsListState,
+    };
+  },
```
Comment (0x009922) [inline src/pages/AccountDetails.vue:96]: Please read the documentation of `useParamScope`. It supports optional scopes:
```suggestion
  () => {
    return (account.value?.owned_assets || false) && {
      key: JSON.stringify(assetsListState),
      payload: assetsListState,
    };
  },
```

File: src/pages/BlockDetails.vue [hunk 17]
```diff
@@ -30,31 +28,31 @@ const blockHeightOrHash = computed(() => {
...
+  ({ payload }) => setupAsyncData(() => http.fetchBlock(payload))
 );
 
+const isBlockLoading = computed(() => blockScope.value.expose.isLoading);
+const block = computed(() => {
+  const res = blockScope.value?.expose.data;
+
+  if (!res) return null;
+
+  return res;
```
Comment (0x009922) [inline src/pages/BlockDetails.vue:47]:
```suggestion
  return blockScope.value?.expose.data ?? null;
```

File: src/pages/DomainDetails.vue [hunk 18]
```diff
@@ -44,35 +42,70 @@ const domainId = computed(() => {
...
 
-const domain = ref<Domain | null>(null);
-const isFetchingDomain = ref(false);
+const domainScope = useParamScope(
+  () => {
+    return {
+      key: domainId.value,
+      payload: domainId.value,
+    };
+  },
```
Comment (0x009922) [inline src/pages/DomainDetails.vue:51]: Primitive payload

File: src/pages/AccountDetails.vue (hunk 16 above)
Comment (VladislavPopovSR) [reply src/pages/AccountDetails.vue:96]: Great! I decided to use `if` condition instead and return either `null` or scope.

File: src/widgets/latest-transactions/LatestTransactions.vue (hunk 13 above)
Comment (VladislavPopovSR) [reply src/widgets/latest-transactions/LatestTransactions.vue]: Forgot to remove :(

**Review by 0x009922 — COMMENTED** [review]
LGTM, but there are some pending conversations

**Review by 0x009922 — APPROVED** [review]

File: package.json (hunk 10 above)
Comment (VladislavPopovSR) [reply package.json:36]: Let's follow newer versions and get away from deprecated functionality
